components/scream/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -0,0 +1,168 @@
+  const SPATimeState& time_state,
+  const SPAPressureState& pressure_state,
+  const SPAData&   data_beg,
+  const SPAData&   data_end,
+  const SPAOutput& data_out,
+  Int ncols,
+  Int nlevs,
+  Int nswbands,
+  Int nlwbands)
+{
+  // Gather time stamp info
+  auto& t_now = time_state.t_now;
+  auto& t_beg = time_state.t_beg_month;
+  auto& t_len = time_state.days_this_month;
+
+  // For now we require that the Data in and the Data out have the same number of columns.
+  EKAT_REQUIRE(ncols==pressure_state.ncols);
+
+  view_2d<Spack> p_src("p_mid_src",ncols,pressure_state.nlevs), 
+                 ccn3_src("ccn3_src",ncols,pressure_state.nlevs);
+  view_3d<Spack> aer_g_sw_src("aer_g_sw_src",ncols,nswbands,pressure_state.nlevs),
+                 aer_ssa_sw_src("aer_ssa_sw_src",ncols,nswbands,pressure_state.nlevs),
+                 aer_tau_sw_src("aer_tau_sw_src",ncols,nswbands,pressure_state.nlevs),
+                 aer_tau_lw_src("aer_tau_lw_src",ncols,nlwbands,pressure_state.nlevs);
+
+  using ExeSpace = typename KT::ExeSpace;
+  const Int nk_pack = ekat::npack<Spack>(nlevs);
+  const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(ncols, nk_pack);
+  // SPA Main loop
+  Kokkos::parallel_for(
+    "spa main loop",
+    policy,
+    KOKKOS_LAMBDA(const MemberType& team) {
+
+    const Int i = team.league_rank();
+
+    // Get single-column subviews of all inputs
+    const auto& ps_beg_sub                 = pressure_state.ps_this_month(i);
+    const auto& ps_end_sub                 = pressure_state.ps_next_month(i);
+    const auto& pmid_sub                   = ekat::subview(pressure_state.pmid, i);
+    const auto& p_src_sub                  = ekat::subview(p_src, i);
+
+    const auto& ccn3_beg_sub               = ekat::subview(data_beg.CCN3, i);
+    const auto& ccn3_end_sub               = ekat::subview(data_end.CCN3, i);
+    const auto& ccn3_src_sub               = ekat::subview(ccn3_src, i);
+
+    // First Step: Horizontal Interpolation if needed - Skip for Now
+  
+    // Second Step: Temporal Interpolation
+    auto slope = (t_now-t_beg)/t_len;
+    /* Determine PS for the source data at this time */
+    auto ps_src  =  ps_beg_sub + slope * (ps_end_sub-ps_beg_sub);
+    /* Reconstruct the vertical pressure profile for the data and time interpolation
+     * of the data */
+    {
+    using C = scream::physics::Constants<Real>;
+    static constexpr auto P0 = C::P0;
+    const Int nk_pack = ekat::npack<Spack>(pressure_state.nlevs);
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nk_pack), [&] (Int k) {
+        p_src_sub(k)    = ps_src * pressure_state.hybm(k) + P0 * pressure_state.hyam(k);
+        ccn3_src_sub(k) = ccn3_beg_sub(k) + slope * (ccn3_end_sub(k) - ccn3_beg_sub(k));
+    });
+    team.team_barrier();
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nk_pack), [&] (int k) {
+      const auto& aer_g_sw_beg_sub           = ekat::subview(data_beg.AER_G_SW, i);
```

Comment:
These subviews don't depend on k, so you could take them _before_ entering the ||for, rather than at every k iteration. Or, alternatively, subview both i and k. The latter is actually not possible as the code is, since it is not slicing the slowest indices. However, read the comment below: if you swap k and n in the two ||fors, then you can slice away column and band index.

Reply:
Yes, thank you for the comment, I was unsure of the most efficient way to set up this parallelism.  I've followed your suggestion and switch the order of the loops.